File: 46-Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0){
            return res;
        }
        
        res.add(new ArrayList<Integer>());
        
        for (int i = 0; i < nums.length; i++){
            List<List<Integer>> currLevel = new ArrayList<>();
            for (int j = 0; j <= i; j++){
                for (List<Integer> l : res){
                    List<Integer> tmp = new ArrayList<>(l);
                    tmp.add(j, nums[i]);
                    currLevel.add(tmp);
                }
            }
            res = currLevel;
        }
        return res;
    }
}

File: cpp/46-Permutations.cpp
/*
[ [1,2,3],
  [1,3,2],
  [2,1,3],
  [2,3,1],
  [3,1,2],
  [3,2,1]]
*/
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> curr;
        if (nums.empty()) {
            return res;
        }
        vector<bool> visited(nums.size(), false);
        permute(res, curr, nums, visited);
        return res;
    }

    void permute(vector<vector<int>>& res, vector<int>& curr, vector<int>& nums, vector<bool>& visited) {
        if (curr.size() == nums.size()) {
            res.push_back(curr);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (visited[i])
                continue;
            curr.push_back(nums[i]);
            visited[i] = true;
            permute(res, curr, nums, visited);
            visited[i] = false;
            curr.pop_back();
        }
    }
};
